File: backend/app/api/sim_requests.py
from datetime import datetime
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import get_current_user, require_approve_requests, _perm
from app.models.user import UserRole
from app.models.sim_request import SimAccessRequest, SimGrant, RequestStatus, PermissionLevel
from app.models.modem import Modem
from app.services import notify

router = APIRouter(prefix="/sim-requests", tags=["sim-requests"])


class RequestCreate(BaseModel):
    modem_id: int
    requested_level: str = "use"
    reason: Optional[str] = None


class ApproveBody(BaseModel):
    granted_level: str = "use"
    expires_at: Optional[datetime] = None
    admin_note: Optional[str] = None


class RejectBody(BaseModel):
    admin_note: Optional[str] = None


class BatchApproveBody(BaseModel):
    ids: List[int]
    granted_level: str = "use"
    expires_at: Optional[datetime] = None
    admin_note: Optional[str] = None


class DirectGrantBody(BaseModel):
    user_id: int
    modem_id: int
    granted_level: str = "use"
    expires_at: Optional[datetime] = None
    admin_note: Optional[str] = None


def _fmt_request(r: SimAccessRequest, grants: dict):
    """Serialize a request, merging current grant info if available."""
    now = datetime.utcnow()
    grant = grants.get((r.user_id, r.modem_id))
    return {
        "id": r.id,
        "user_id": r.user_id,
        "username": r.user.username if r.user else None,
        "modem_id": r.modem_id,
        "modem_name": (r.modem.alias or f"SIM {r.modem_id}") if r.modem else f"SIM {r.modem_id}",
        "status": r.status,
        "requested_level": r.requested_level,
        "granted_level": grant.granted_level if grant else None,
        "reason": r.reason,
        "admin_note": r.admin_note,
        "expires_at": grant.expires_at.isoformat() if grant and grant.expires_at else None,
        "created_at": r.created_at.isoformat() if r.created_at else None,
        "updated_at": r.updated_at.isoformat() if r.updated_at else None,
        "is_expired": bool(grant and grant.expires_at and grant.expires_at < now),
    }


def _fmt_grant(g: SimGrant):
    now = datetime.utcnow()
    return {
        "id": g.id,
        "user_id": g.user_id,
        "modem_id": g.modem_id,
        "granted_level": g.granted_level,
        "expires_at": g.expires_at.isoformat() if g.expires_at else None,
        "is_expired": bool(g.expires_at and g.expires_at < now),
        "created_at": g.created_at.isoformat() if g.created_at else None,
    }


def _approver_modem_scope(approver, db: Session) -> Optional[List[int]]:
    if approver.role == UserRole.ADMIN:
        return None
    p = _perm(approver)
    return p.get("allowed_modem_ids") if p else []


def _upsert_grant(db: Session, user_id: int, modem_id: int,
                  granted_level: str, expires_at, granted_by_id: int,
                  request_id: Optional[int] = None):
    """Insert or update sim_grants for (user_id, modem_id)."""
    now = datetime.utcnow()
    existing = db.query(SimGrant).filter(
        SimGrant.user_id == user_id,
        SimGrant.modem_id == modem_id,
    ).first()
    if existing:
        existing.granted_level = PermissionLevel(granted_level)
        existing.expires_at = expires_at
        existing.granted_by_id = granted_by_id
        if request_id:
            existing.request_id = request_id
        existing.updated_at = now
    else:
        db.add(SimGrant(
            user_id=user_id,
            modem_id=modem_id,
            granted_level=PermissionLevel(granted_level),
            expires_at=expires_at,
            granted_by_id=granted_by_id,
            request_id=request_id,
            created_at=now,
            updated_at=now,
        ))


# ── Requests ───────────────────────────────────────────────────────────────────

@router.post("/")
def create_request(body: RequestCreate, db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    if body.requested_level not in ("view", "use"):
        raise HTTPException(400, "requested_level 必须是 view 或 use")

    # Block if already has active grant
    now = datetime.utcnow()
    existing_grant = db.query(SimGrant).filter(
        SimGrant.user_id == current_user.id,
        SimGrant.modem_id == body.modem_id,
    ).first()
    if existing_grant and (existing_grant.expires_at is None or existing_grant.expires_at > now):
        raise HTTPException(400, "已有有效授权，无需重复申请")

    # Block if already pending
    pending = db.query(SimAccessRequest).filter(
        SimAccessRequest.user_id == current_user.id,
        SimAccessRequest.modem_id == body.modem_id,
        SimAccessRequest.status == RequestStatus.PENDING,
    ).first()
    if pending:
        raise HTTPException(400, "已有待审批的申请，请勿重复提交")

    req = SimAccessRequest(
        user_id=current_user.id,
        modem_id=body.modem_id,
        requested_level=PermissionLevel(body.requested_level),
        reason=body.reason,
        status=RequestStatus.PENDING,
    )
    db.add(req)
    db.commit()
    notify.push("sim_request", "新的SIM卡申请",
                f"用户 {current_user.username} 申请访问 SIM {body.modem_id}",
                audience="admin")
    return {"ok": True}


@router.get("/my")
def my_requests(db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    reqs = (
        db.query(SimAccessRequest)
        .filter(SimAccessRequest.user_id == current_user.id)
        .order_by(SimAccessRequest.created_at.desc())
        .all()
    )
    grants = {(g.user_id, g.modem_id): g for g in
              db.query(SimGrant).filter(SimGrant.user_id == current_user.id).all()}
    return [_fmt_request(r, grants) for r in reqs]


@router.get("/my-grants")
def my_grants(db: Session = Depends(get_db), current_user=Depends(get_current_user)):
    """Return the user's current active grants."""
    now = datetime.utcnow()
    grants = db.query(SimGrant).filter(
        SimGrant.user_id == current_user.id,
    ).all()
    return [_fmt_grant(g) for g in grants if g.expires_at is None or g.expires_at > now]


@router.get("/")
def list_requests(
    status: Optional[str] = None,
    db: Session = Depends(get_db),
    approver=Depends(require_approve_requests),
):
    scope = _approver_modem_scope(approver, db)
    q = db.query(SimAccessRequest)
    if scope is not None:
        q = q.filter(SimAccessRequest.modem_id.in_(scope))
    if status:
        q = q.filter(SimAccessRequest.status == status)
    reqs = q.order_by(SimAccessRequest.created_at.desc()).all()

    user_ids = {r.user_id for r in reqs}
    modem_ids = {r.modem_id for r in reqs}
    grants = {}
    if user_ids and modem_ids:
        for g in db.query(SimGrant).filter(
            SimGrant.user_id.in_(user_ids),
            SimGrant.modem_id.in_(modem_ids),
        ).all():
            grants[(g.user_id, g.modem_id)] = g

    return [_fmt_request(r, grants) for r in reqs]


@router.put("/{req_id}/approve")
def approve_request(req_id: int, body: ApproveBody, db: Session = Depends(get_db), approver=Depends(require_approve_requests)):
    if body.granted_level not in ("view", "use"):
        raise HTTPException(400, "granted_level 必须是 view 或 use")

    req = db.query(SimAccessRequest).filter(SimAccessRequest.id == req_id).first()
    if not req:
        raise HTTPException(404, "申请不存在")

    scope = _approver_modem_scope(approver, db)
    if scope is not None and req.modem_id not in scope:
        raise HTTPException(403, "无权审批该设备的申请")

    req.status = RequestStatus.APPROVED
    req.admin_note = body.admin_note
    req.updated_at = datetime.utcnow()

    _upsert_grant(db, req.user_id, req.modem_id, body.granted_level,
                  body.expires_at, approver.id, req.id)
    db.commit()

    modem = db.query(Modem).filter(Modem.id == req.modem_id).first()
    modem_name = (modem.alias or f"SIM {req.modem_id}") if modem else f"SIM {req.modem_id}"
    level_label = "使用权限" if body.granted_level == "use" else "查看权限"
    exp_str = f"，有效期至 {body.expires_at.strftime('%Y-%m-%d')}" if body.expires_at else "（永久）"
    notify.push("sim_approved", "SIM卡申请已批准",
                f"你对 {modem_name} 的申请已获批准{level_label}{exp_str}",
                audience="user", target_user_id=req.user_id)
    return {"ok": True}


@router.put("/{req_id}/reject")
def reject_request(req_id: int, body: RejectBody, db: Session = Depends(get_db), approver=Depends(require_approve_requests)):
    req = db.query(SimAccessRequest).filter(SimAccessRequest.id == req_id).first()
    if not req:
        raise HTTPException(404, "申请不存在")

    scope = _approver_modem_scope(approver, db)
    if scope is not None and req.modem_id not in scope:
        raise HTTPException(403, "无权审批该设备的申请")

    req.status = RequestStatus.REJECTED
    req.admin_note = body.admin_note
    req.updated_at = datetime.utcnow()
    db.commit()

    modem = db.query(Modem).filter(Modem.id == req.modem_id).first()
    modem_name = (modem.alias or f"SIM {req.modem_id}") if modem else f"SIM {req.modem_id}"
    notify.push("sim_rejected", "SIM卡申请已拒绝",
                f"你对 {modem_name} 的申请未获批准" + (f"，原因：{body.admin_note}" if body.admin_note else ""),
                audience="user", target_user_id=req.user_id)
    return {"ok": True}


@router.post("/batch-approve")
def batch_approve(body: BatchApproveBody, db: Session = Depends(get_db), approver=Depends(require_approve_requests)):
    if body.granted_level not in ("view", "use"):
        raise HTTPException(400, "granted_level 必须是 view 或 use")

    scope = _approver_modem_scope(approver, db)
    reqs = db.query(SimAccessRequest).filter(SimAccessRequest.id.in_(body.ids)).all()
    approved_count = 0
    for req in reqs:
        if scope is not None and req.modem_id not in scope:
            continue
        req.status = RequestStatus.APPROVED
        req.admin_note = body.admin_note
        req.updated_at = datetime.utcnow()
        _upsert_grant(db, req.user_id, req.modem_id, body.granted_level,
                      body.expires_at, approver.id, req.id)
        modem = db.query(Modem).filter(Modem.id == req.modem_id).first()
        modem_name = (modem.alias or f"SIM {req.modem_id}") if modem else f"SIM {req.modem_id}"
        level_label = "使用权限" if body.granted_level == "use" else "查看权限"
        exp_str = f"，有效期至 {body.expires_at.strftime('%Y-%m-%d')}" if body.expires_at else "（永久）"
        notify.push("sim_approved", "SIM卡申请已批准",
                    f"你对 {modem_name} 的申请已获批准{level_label}{exp_str}",
                    audience="user", target_user_id=req.user_id)
        approved_count += 1
    db.commit()
    return {"approved": approved_count}


@router.post("/grant")
def direct_grant(body: DirectGrantBody, db: Session = Depends(get_db), approver=Depends(require_approve_requests)):
    """Approver directly grants access without requiring a prior application."""
    if body.granted_level not in ("view", "use"):
        raise HTTPException(400, "granted_level 必须是 view 或 use")

    scope = _approver_modem_scope(approver, db)
    if scope is not None and body.modem_id not in scope:
        raise HTTPException(403, "无权授权该设备")

    modem = db.query(Modem).filter(Modem.id == body.modem_id).first()
    if not modem:
        raise HTTPException(404, "设备不存在")

    _upsert_grant(db, body.user_id, body.modem_id, body.granted_level,
                  body.expires_at, approver.id)
    db.commit()

    modem_name = modem.alias or f"SIM {modem.id}"
    level_label = "使用权限" if body.granted_level == "use" else "查看权限"
    notify.push("sim_approved", "SIM卡权限已授予",
                f"管理员已授予你 {modem_name} 的{level_label}",
                audience="user", target_user_id=body.user_id)
    return {"ok": True}


@router.delete("/grants/{grant_id}")
def revoke_grant(grant_id: int, db: Session = Depends(get_db), approver=Depends(require_approve_requests)):
    """Revoke an active grant."""
    grant = db.query(SimGrant).filter(SimGrant.id == grant_id).first()
    if not grant:
        raise HTTPException(404, "授权记录不存在")
    scope = _approver_modem_scope(approver, db)
    if scope is not None and grant.modem_id not in scope:
        raise HTTPException(403, "无权撤销该设备的授权")
    db.delete(grant)
    db.commit()
    notify.push("sim_revoked", "SIM卡权限已撤销",
                f"你对 SIM {grant.modem_id} 的访问权限已被撤销",
                audience="user", target_user_id=grant.user_id)
    return {"ok": True}

File: backend-go/handlers/sim_requests.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"simnexus-go/database"
	"simnexus-go/middleware"
	"simnexus-go/models"
	"simnexus-go/services"

	"github.com/gin-gonic/gin"
)

// approverModemScope 返回审批员可管理的设备 ID 集合。
// unrestricted=true 表示无限制（管理员或 ModemScope 为空的审批角色）。
func approverModemScope(u *models.User) ([]uint, bool) {
	if u.IsAdmin() {
		return nil, true
	}
	var approverRoles []models.Role
	for _, r := range u.RbacRoles {
		if r.CanApproveRequests {
			approverRoles = append(approverRoles, r)
		}
	}
	if len(approverRoles) == 0 {
		return []uint{}, false
	}
	set := map[uint]struct{}{}
	for _, r := range approverRoles {
		if len(r.ModemScope) == 0 {
			return nil, true
		}
		for _, m := range r.ModemScope {
			set[m.ID] = struct{}{}
		}
	}
	ids := make([]uint, 0, len(set))
	for id := range set {
		ids = append(ids, id)
	}
	return ids, false
}

// inScope 判断给定设备是否在审批员的管辖范围内。
func inScope(ids []uint, unrestricted bool, modemID uint) bool {
	if unrestricted {
		return true
	}
	for _, id := range ids {
		if id == modemID {
			return true
		}
	}
	return false
}

// fmtRequest 将申请记录格式化为 API 响应，包含用户名、设备名、授权状态和是否过期。
func fmtRequest(r *models.SimAccessRequest, grants map[[2]uint]*models.SimGrant) gin.H {
	now := time.Now()
	var username, modemName string
	var user models.User
	if database.DB.First(&user, r.UserID).Error == nil {
		username = user.Username
	}
	var modem models.Modem
	if database.DB.First(&modem, r.ModemID).Error == nil && modem.Alias != "" {
		modemName = modem.Alias
	} else {
		modemName = "SIM " + strconv.Itoa(int(r.ModemID))
	}
	g := grants[[2]uint{r.UserID, r.ModemID}]
	var grantedLevel interface{}
	var expiresAt interface{}
	isExpired := false
	if g != nil {
		grantedLevel = g.GrantedLevel
		if g.ExpiresAt != nil {
			expiresAt = g.ExpiresAt.Format(time.RFC3339)
			isExpired = g.ExpiresAt.Before(now)
		}
	}
	return gin.H{
		"id":              r.ID,
		"user_id":         r.UserID,
		"username":        username,
		"modem_id":        r.ModemID,
		"modem_name":      modemName,
		"status":          r.Status,
		"requested_level": r.RequestedLevel,
		"granted_level":   grantedLevel,
		"reason":          r.Reason,
		"admin_note":      r.AdminNote,
		"expires_at":      expiresAt,
		"created_at":      r.CreatedAt,
		"updated_at":      r.UpdatedAt,
		"is_expired":      isExpired,
	}
}

// upsertGrant 创建或更新 (userID, modemID) 的授权记录，已有记录则原地更新。
func upsertGrant(userID, modemID uint, level string, expiresAt *time.Time, grantedByID uint, requestID *uint) {
	now := time.Now()
	var existing models.SimGrant
	err := database.DB.Where("user_id = ? AND modem_id = ?", userID, modemID).First(&existing).Error
	if err == nil {
		existing.GrantedLevel = level
		existing.ExpiresAt = expiresAt
		existing.GrantedByID = &grantedByID
		if requestID != nil {
			existing.RequestID = requestID
		}
		existing.UpdatedAt = now
		database.DB.Save(&existing)
	} else {
		database.DB.Create(&models.SimGrant{
			UserID: userID, ModemID: modemID, GrantedLevel: level,
			ExpiresAt: expiresAt, GrantedByID: &grantedByID, RequestID: requestID,
			CreatedAt: now, UpdatedAt: now,
		})
	}
}

// modemName 返回设备的展示名称（别名优先，否则为 "SIM <id>"）。
func modemName(modemID uint) string {
	var modem models.Modem
	if database.DB.First(&modem, modemID).Error == nil && modem.Alias != "" {
		return modem.Alias
	}
	return "SIM " + strconv.Itoa(int(modemID))
}

type requestCreate struct {
	ModemID        uint   `json:"modem_id"`
	RequestedLevel string `json:"requested_level"`
	Reason         string `json:"reason"`
}

// CreateSimRequest submits an access request.
func CreateSimRequest(c *gin.Context) {
	me := middleware.CurrentUser(c)
	var body requestCreate
	c.ShouldBindJSON(&body)
	if body.RequestedLevel == "" {
		body.RequestedLevel = models.LevelUse
	}
	if body.RequestedLevel != models.LevelView && body.RequestedLevel != models.LevelUse {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "requested_level 必须是 view 或 use"})
		return
	}
	now := time.Now()
	var eg models.SimGrant
	if database.DB.Where("user_id = ? AND modem_id = ?", me.ID, body.ModemID).First(&eg).Error == nil {
		if eg.ExpiresAt == nil || eg.ExpiresAt.After(now) {
			c.JSON(http.StatusBadRequest, gin.H{"detail": "已有有效授权，无需重复申请"})
			return
		}
	}
	var pending models.SimAccessRequest
	if database.DB.Where("user_id = ? AND modem_id = ? AND status = ?", me.ID, body.ModemID, models.ReqPending).First(&pending).Error == nil {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "已有待审批的申请，请勿重复提交"})
		return
	}
	var reason *string
	if body.Reason != "" {
		reason = &body.Reason
	}
	req := models.SimAccessRequest{
		UserID: me.ID, ModemID: body.ModemID,
		RequestedLevel: body.RequestedLevel, Reason: reason, Status: models.ReqPending,
	}
	database.DB.Create(&req)
	services.Push("sim_request", "新的SIM卡申请",
		"用户 "+me.Username+" 申请访问 SIM "+strconv.Itoa(int(body.ModemID)), "admin", nil)
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

// MyRequests returns the user's requests.
func MyRequests(c *gin.Context) {
	me := middleware.CurrentUser(c)
	var reqs []models.SimAccessRequest
	database.DB.Where("user_id = ?", me.ID).Order("created_at desc").Find(&reqs)
	var grants []models.SimGrant
	database.DB.Where("user_id = ?", me.ID).Find(&grants)
	gm := map[[2]uint]*models.SimGrant{}
	for i := range grants {
		gm[[2]uint{grants[i].UserID, grants[i].ModemID}] = &grants[i]
	}
	out := make([]gin.H, 0, len(reqs))
	for i := range reqs {
		out = append(out, fmtRequest(&reqs[i], gm))
	}
	c.JSON(http.StatusOK, out)
}

// MyGrants returns active grants for the user.
func MyGrants(c *gin.Context) {
	me := middleware.CurrentUser(c)
	now := time.Now()
	var grants []models.SimGrant
	database.DB.Where("user_id = ?", me.ID).Find(&grants)
	out := make([]gin.H, 0)
	for _, g := range grants {
		if g.ExpiresAt != nil && g.ExpiresAt.Before(now) {
			continue
		}
		var exp interface{}
		if g.ExpiresAt != nil {
			exp = g.ExpiresAt.Format(time.RFC3339)
		}
		out = append(out, gin.H{
			"id": g.ID, "user_id": g.UserID, "modem_id": g.ModemID,
			"granted_level": g.GrantedLevel, "expires_at": exp,
			"is_expired": false, "created_at": g.CreatedAt,
		})
	}
	c.JSON(http.StatusOK, out)
}

// ListRequests returns requests within the approver's scope.
func ListRequests(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	ids, unrestricted := approverModemScope(approver)
	q := database.DB.Model(&models.SimAccessRequest{})
	if !unrestricted {
		q = q.Where("modem_id IN ?", ids)
	}
	if st := c.Query("status"); st != "" {
		q = q.Where("status = ?", st)
	}
	var reqs []models.SimAccessRequest
	q.Order("created_at desc").Find(&reqs)

	gm := map[[2]uint]*models.SimGrant{}
	if len(reqs) > 0 {
		userIDs := map[uint]bool{}
		modemIDs := map[uint]bool{}
		for _, r := range reqs {
			userIDs[r.UserID] = true
			modemIDs[r.ModemID] = true
		}
		var grants []models.SimGrant
		database.DB.Where("user_id IN ? AND modem_id IN ?", keys(userIDs), keys(modemIDs)).Find(&grants)
		for i := range grants {
			gm[[2]uint{grants[i].UserID, grants[i].ModemID}] = &grants[i]
		}
	}
	out := make([]gin.H, 0, len(reqs))
	for i := range reqs {
		out = append(out, fmtRequest(&reqs[i], gm))
	}
	c.JSON(http.StatusOK, out)
}

type approveBody struct {
	GrantedLevel string     `json:"granted_level"`
	ExpiresAt    *time.Time `json:"expires_at"`
	AdminNote    string     `json:"admin_note"`
}

// ApproveRequest approves a single request.
func ApproveRequest(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	id, _ := strconv.Atoi(c.Param("id"))
	var body approveBody
	c.ShouldBindJSON(&body)
	if body.GrantedLevel == "" {
		body.GrantedLevel = models.LevelUse
	}
	if body.GrantedLevel != models.LevelView && body.GrantedLevel != models.LevelUse {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "granted_level 必须是 view 或 use"})
		return
	}
	var req models.SimAccessRequest
	if database.DB.First(&req, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "申请不存在"})
		return
	}
	ids, unrestricted := approverModemScope(approver)
	if !inScope(ids, unrestricted, req.ModemID) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权审批该设备的申请"})
		return
	}
	req.Status = models.ReqApproved
	setNote(&req, body.AdminNote)
	req.UpdatedAt = time.Now()
	database.DB.Save(&req)
	upsertGrant(req.UserID, req.ModemID, body.GrantedLevel, body.ExpiresAt, approver.ID, &req.ID)
	notifyApproved(req.UserID, req.ModemID, body.GrantedLevel, body.ExpiresAt)
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

type rejectBody struct {
	AdminNote string `json:"admin_note"`
}

// RejectRequest rejects a request.
func RejectRequest(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	id, _ := strconv.Atoi(c.Param("id"))
	var body rejectBody
	c.ShouldBindJSON(&body)
	var req models.SimAccessRequest
	if database.DB.First(&req, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "申请不存在"})
		return
	}
	ids, unrestricted := approverModemScope(approver)
	if !inScope(ids, unrestricted, req.ModemID) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权审批该设备的申请"})
		return
	}
	req.Status = models.ReqRejected
	setNote(&req, body.AdminNote)
	req.UpdatedAt = time.Now()
	database.DB.Save(&req)
	body2 := "你对 " + modemName(req.ModemID) + " 的申请未获批准"
	if body.AdminNote != "" {
		body2 += "，原因：" + body.AdminNote
	}
	services.Push("sim_rejected", "SIM卡申请已拒绝", body2, "user", &req.UserID)
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

type batchApproveBody struct {
	IDs          []uint     `json:"ids"`
	GrantedLevel string     `json:"granted_level"`
	ExpiresAt    *time.Time `json:"expires_at"`
	AdminNote    string     `json:"admin_note"`
}

// BatchApprove approves multiple requests.
func BatchApprove(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	var body batchApproveBody
	c.ShouldBindJSON(&body)
	if body.GrantedLevel == "" {
		body.GrantedLevel = models.LevelUse
	}
	if body.GrantedLevel != models.LevelView && body.GrantedLevel != models.LevelUse {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "granted_level 必须是 view 或 use"})
		return
	}
	ids, unrestricted := approverModemScope(approver)
	var reqs []models.SimAccessRequest
	database.DB.Where("id IN ?", body.IDs).Find(&reqs)
	count := 0
	for i := range reqs {
		req := &reqs[i]
		if !inScope(ids, unrestricted, req.ModemID) {
			continue
		}
		req.Status = models.ReqApproved
		setNote(req, body.AdminNote)
		req.UpdatedAt = time.Now()
		database.DB.Save(req)
		upsertGrant(req.UserID, req.ModemID, body.GrantedLevel, body.ExpiresAt, approver.ID, &req.ID)
		notifyApproved(req.UserID, req.ModemID, body.GrantedLevel, body.ExpiresAt)
		count++
	}
	c.JSON(http.StatusOK, gin.H{"approved": count})
}

type directGrantBody struct {
	UserID       uint       `json:"user_id"`
	ModemID      uint       `json:"modem_id"`
	GrantedLevel string     `json:"granted_level"`
	ExpiresAt    *time.Time `json:"expires_at"`
	AdminNote    string     `json:"admin_note"`
}

// DirectGrant grants access without a prior request.
func DirectGrant(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	var body directGrantBody
	c.ShouldBindJSON(&body)
	if body.GrantedLevel == "" {
		body.GrantedLevel = models.LevelUse
	}
	if body.GrantedLevel != models.LevelView && body.GrantedLevel != models.LevelUse {
		c.JSON(http.StatusBadRequest, gin.H{"detail": "granted_level 必须是 view 或 use"})
		return
	}
	ids, unrestricted := approverModemScope(approver)
	if !inScope(ids, unrestricted, body.ModemID) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权授权该设备"})
		return
	}
	var modem models.Modem
	if database.DB.First(&modem, body.ModemID).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "设备不存在"})
		return
	}
	upsertGrant(body.UserID, body.ModemID, body.GrantedLevel, body.ExpiresAt, approver.ID, nil)
	label := "使用权限"
	if body.GrantedLevel == models.LevelView {
		label = "查看权限"
	}
	services.Push("sim_approved", "SIM卡权限已授予",
		"管理员已授予你 "+modemName(body.ModemID)+" 的"+label, "user", &body.UserID)
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

// RevokeGrant removes an active grant.
func RevokeGrant(c *gin.Context) {
	approver := middleware.CurrentUser(c)
	id, _ := strconv.Atoi(c.Param("id"))
	var grant models.SimGrant
	if database.DB.First(&grant, id).Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"detail": "授权记录不存在"})
		return
	}
	ids, unrestricted := approverModemScope(approver)
	if !inScope(ids, unrestricted, grant.ModemID) {
		c.JSON(http.StatusForbidden, gin.H{"detail": "无权撤销该设备的授权"})
		return
	}
	uid := grant.UserID
	mid := grant.ModemID
	database.DB.Delete(&grant)
	services.Push("sim_revoked", "SIM卡权限已撤销",
		"你对 SIM "+strconv.Itoa(int(mid))+" 的访问权限已被撤销", "user", &uid)
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

// setNote 将审批备注写入申请记录，空字符串时置 nil。
func setNote(r *models.SimAccessRequest, note string) {
	if note == "" {
		r.AdminNote = nil
	} else {
		r.AdminNote = &note
	}
}

// notifyApproved 向用户推送申请批准通知，包含权限级别和有效期信息。
func notifyApproved(userID, modemID uint, level string, expiresAt *time.Time) {
	levelLabel := "使用权限"
	if level == models.LevelView {
		levelLabel = "查看权限"
	}
	expStr := "（永久）"
	if expiresAt != nil {
		expStr = "，有效期至 " + expiresAt.Format("2006-01-02")
	}
	services.Push("sim_approved", "SIM卡申请已批准",
		"你对 "+modemName(modemID)+" 的申请已获批准"+levelLabel+expStr, "user", &userID)
}

// keys 将 map[uint]bool 的所有键提取为 slice。
func keys(m map[uint]bool) []uint {
	out := make([]uint, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}
